### PR TITLE
AIP-83: Restore Uniqueness Constraint on Logical Date, Make It Nullable

### PR DESCRIPTION
**Pull request by vatsrahul1001**
closes:https://github.com/apache/airflow/issues/46187

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by vatsrahul1001** [comment]
Need to look at failures.

File: tests/api_fastapi/core_api/routes/public/test_assets.py [hunk 1]
```diff
@@ -149,7 +149,7 @@ def _create_dag_run(session, num: int = 2):
             dag_id="source_dag_id",
             run_id=f"source_run_id_{i}",
             run_type=DagRunType.MANUAL,
-            logical_date=DEFAULT_DATE,
+            logical_date=DEFAULT_DATE + timedelta(days=i),
```
Comment (Lee-W) [inline tests/api_fastapi/core_api/routes/public/test_assets.py]: Why did we change these dates here?

File: tests/api_fastapi/core_api/routes/public/test_dag_run.py [hunk 2]
```diff
@@ -1333,28 +1332,8 @@ def test_should_response_200_for_duplicate_logical_date(self, test_client):
...
-                "last_scheduling_decision": None,
-                "run_type": "manual",
-                "state": "queued",
-                "external_trigger": True,
-                "triggered_by": "rest_api",
-                "conf": {},
-                "note": note,
-            }
+        assert response_1.status_code == 200
+        assert response_2.status_code == 409
```
Comment (Lee-W) [inline tests/api_fastapi/core_api/routes/public/test_dag_run.py:1355]: should we keep the body check?

File: airflow/models/backfill.py [hunk 3]
```diff
@@ -262,7 +266,23 @@ def _create_backfill_dag_run(
     dag_run_conf,
     backfill_sort_ordinal,
     session,
+    from_date,
+    to_date,
```
Comment (Lee-W) [inline airflow/models/backfill.py]: should we add type annotation here?

File: tests/api_fastapi/core_api/routes/public/test_task_instances.py [hunk 4]
```diff
@@ -1027,20 +1028,26 @@ def test_should_respond_200_for_dag_id_filter(self, test_client, session):
         assert count == len(response.json()["task_instances"])
 
     @pytest.mark.parametrize(
-        "order_by_field", ["start_date", "logical_date", "data_interval_start", "data_interval_end"]
+        "order_by_field, date",
```
Comment (Lee-W) [inline tests/api_fastapi/core_api/routes/public/test_task_instances.py]:
```suggestion
        "order_by_field, base_date",
```
base_date might be better. `date` might shadow `from datetime import date`

File: tests/models/test_dagrun.py [hunk 5]
```diff
@@ -355,10 +355,11 @@ def test_dagrun_no_deadlock_with_depends_on_past(self, dag_maker, session):
             run_id="test_dagrun_no_deadlock_1",
             start_date=DEFAULT_DATE,
         )
-        dr2 = dag_maker.create_dagrun_after(
-            dr,
+        next_date = DEFAULT_DATE + datetime.timedelta(days=1)
+        dr2 = dag_maker.create_dagrun(
             run_id="test_dagrun_no_deadlock_2",
             start_date=DEFAULT_DATE + datetime.timedelta(days=1),
+            logical_date=next_date,
```
Comment (Lee-W) [inline tests/models/test_dagrun.py]: start_date seems to use this next_date as well so we can just reuse it and would like to know why do we set it to next_date here any date equal or after start_date?

File: tests/api_fastapi/core_api/routes/public/test_assets.py (hunk 1 above)
Comment (vatsrahul1001) [reply tests/api_fastapi/core_api/routes/public/test_assets.py]: This test was failing due to unique constraint error as same logical date was being inserted in DAGRUN

Comment (Lee-W) [reply tests/api_fastapi/core_api/routes/public/test_assets.py]: got it. makes sense

File: tests/api_fastapi/core_api/routes/public/test_task_instances.py (hunk 4 above)
Comment (vatsrahul1001) [reply tests/api_fastapi/core_api/routes/public/test_task_instances.py]: Implemented change

File: airflow/models/backfill.py (hunk 3 above)
Comment (vatsrahul1001) [reply airflow/models/backfill.py]: Implemented change

File: tests/api_fastapi/core_api/routes/public/test_dag_run.py (hunk 2 above)
Comment (vatsrahul1001) [reply tests/api_fastapi/core_api/routes/public/test_dag_run.py:1355]: Added body check for 200

File: tests/models/test_dagrun.py (hunk 5 above)
Comment (vatsrahul1001) [reply tests/models/test_dagrun.py]: Implemented change

File: airflow/models/backfill.py [hunk 6]
```diff
@@ -194,7 +194,11 @@ def _get_latest_dag_run_row_query(info, session):
 def _get_dag_run_no_create_reason(dr, reprocess_behavior: ReprocessBehavior) -> str | None:
     non_create_reason = None
     if dr.state not in (DagRunState.SUCCESS, DagRunState.FAILED):
-        non_create_reason = BackfillDagRunExceptionReason.IN_FLIGHT
+        if dr.clear_number == 0:
+            non_create_reason = BackfillDagRunExceptionReason.IN_FLIGHT
+        else:
+            non_create_reason = BackfillDagRunExceptionReason.CLEARED_RUN
```
Comment (dstandish) [inline airflow/models/backfill.py]: isn't it still true that it is in flight, in this case? it might also be true that it was cleared; but, the primary reason would still seem, to me, to be that the run is in flight

Comment (vatsrahul1001) [reply airflow/models/backfill.py]: Do you mean we do not need a CLEARED_RUN state and instead we can use IN_FLIGHT for a clear run as well?

File: airflow/models/backfill.py [hunk 7]
```diff
@@ -262,7 +266,23 @@ def _create_backfill_dag_run(
     dag_run_conf,
     backfill_sort_ordinal,
     session,
+    from_date: datetime,
+    to_date: datetime,
```
Comment (dstandish) [inline airflow/models/backfill.py]: this function is supposed to create only a single dag run.  so this does not seem right. you clear a range of dag runs below.  and if you're clearing, you're not creating.

**Review by dstandish — COMMENTED** [review]
@vatsrahul1001 left a comment re how you're doing the clearing. it think that needs more attention.

it might be better also to split this into two PRs.  one where you change the behavior of backfill to clear the run when there's an existing one.  and the other to restore uniqueness.  probably uniqueness would have to come after the backfill change since it would break tests.

**Review by Lee-W — COMMENTED** [review]
Aside from Daniel's comments, everything looks good to me.

**Comment by uranusjr** [comment]
I added a couple of changes to fix tests in ways I think make more sense. This looks good to me except the parts involving backfill; deferring those to Daniel.

File: tests/models/test_dagrun.py [hunk 8]
```diff
@@ -353,6 +353,7 @@ def test_dagrun_no_deadlock_with_depends_on_past(self, dag_maker, session):
```
Comment (sunank200) [inline tests/models/test_dagrun.py:353]: I think we should add one more test for `logical_date` as null in this file.

**Review by sunank200 — COMMENTED** [review]
Apart from the above comments I think we should add more test for logical_dates as null especially in models.

File: airflow/models/backfill.py (hunk 7 above)
Comment (uranusjr) [reply airflow/models/backfill.py]: I changed this to use run_id to clear one specific run. I’m not sure about the condition when to do the clear (only when the run is not currently running, and under specific reprocess behaviors), so that is left unchanged.

Comment (vatsrahul1001) [reply airflow/models/backfill.py]: @uranusjr I tested this with run_id change and looks like run is not getting cleared.

**Comment by dstandish** [comment]
Let's take the backfill part of this out; we can just do minimal changes to backfill such that tests are passing.  Backfill can be broken if necessary as a result of this and fixed by sneha's PR.   But we need to get the constraint change in there.

**Comment by uranusjr** [comment]
Let’s take out the changes in backfill.py from this PR so we can merge the model changes first. Everything else sort of depends on them. This means backfilling would break if you hit a duplicate logical date, but I think it’s fine as long as we fix it before the final release.

I’ll do that tomorrow if nobody beats me to it.

**Comment by vatsrahul1001** [comment]
@dstandish @uranusjr Earlier, I had opened [PR #46232](https://github.com/apache/airflow/pull/46232), which only contained logical date-related changes. However, as discussed with @Lee-W, we closed that PR because the backfill reprocess test was failing. Should I reopen that PR or remove the backfill-related changes from this one?

**Comment by vatsrahul1001** [comment]
@uranusjr @dstandish I have removed backfill related changes from this PR. We should be good to merge this after CI is green

File: airflow/models/backfill.py (hunk 7 above)
Comment (vatsrahul1001) [reply airflow/models/backfill.py]: Backfill changes will be done in separate PR

File: airflow/models/backfill.py (hunk 6 above)
Comment (vatsrahul1001) [reply airflow/models/backfill.py]: Backfill changes will be done in seprate PR

**Comment by dstandish** [comment]
> @uranusjr @dstandish I have removed backfill related changes from this PR. We should be good to merge this after CI is green

ok cool

File: airflow/migrations/versions/0032_3_0_0_rename_execution_date_to_logical_date_and_nullable.py [hunk 9]
```diff
@@ -61,9 +66,11 @@ def downgrade():
             "logical_date",
             new_column_name="execution_date",
             existing_type=TIMESTAMP(timezone=True),
-            existing_nullable=False,
+            nullable=False,
```
Comment (dstandish) [inline airflow/migrations/versions/0032_3_0_0_rename_execution_date_to_logical_date_and_nullable.py:69]: in the downgrade, we would need to do something about the nulls. either delete the runs, or move them to an archive table, or populate them with 1900-01-01 or something.  @jedcunningham ?

File: airflow/models/dagrun.py [hunk 10]
```diff
@@ -1321,8 +1322,12 @@ def verify_integrity(self, *, session: Session = NEW_SESSION) -> None:
         def task_filter(task: Operator) -> bool:
             return task.task_id not in task_ids and (
                 self.run_type == DagRunType.BACKFILL_JOB
```
Comment (dstandish) [inline airflow/models/dagrun.py:1324]: unrelated, by why are we filtering out backfill jobs?  that's weird.

Comment (dstandish) [reply airflow/models/dagrun.py:1324]: oh, this means we ignore task start date / end date when running backfill

File: airflow/migrations/versions/0032_3_0_0_rename_execution_date_to_logical_date_and_nullable.py (hunk 9 above)
Comment (jedcunningham) [reply airflow/migrations/versions/0032_3_0_0_rename_execution_date_to_logical_date_and_nullable.py:69]: hmm, depends how cheeky I'm feeling haha. The downside of populating a old fake value is the need for it to be unique, seems like a pain.

**Review by Lee-W — APPROVED** [review]
LGTM. @uranusjr if it also looks good to me, I think we could probably just merge it?

File: airflow/migrations/versions/0032_3_0_0_rename_execution_date_to_logical_date_and_nullable.py (hunk 9 above)
Comment (uranusjr) [reply airflow/migrations/versions/0032_3_0_0_rename_execution_date_to_logical_date_and_nullable.py:69]: Maybe move them to a separate table? We did that for some records when we added many unique constraints on TI back in 2.2 or something.

**Review by uranusjr — APPROVED** [review]
Made a change in tests, I’ll merge this once everything is green.

File: airflow/migrations/versions/0032_3_0_0_rename_execution_date_to_logical_date_and_nullable.py (hunk 9 above)
Comment (uranusjr) [reply airflow/migrations/versions/0032_3_0_0_rename_execution_date_to_logical_date_and_nullable.py:69]: I created #46500 for this.

**Comment by potiuk** [comment]
> It's curious/a massive oversight in tests that this is broken but it's only a provider test that picks it up!

Yep. I would really be curious to see what's the solution - maybe it's just a test case, maybe that's something that it's missing in the unit tests. Or maybe we should bring back all provider tests to be run on core changes - they are disabled now but seems to be breaking things more often.
